src/libraries/Microsoft.Extensions.Configuration.Binder/src/ConfigurationBinder.cs

```diff
@@ -589,12 +592,10 @@ private static bool CanBindToTheseConstructorParameters(ParameterInfo[] construc
 
             Debug.Assert(dictionary is not null);
 
-            Type dictionaryObjectType = dictionary.GetType();
 
-            MethodInfo tryGetValue = dictionaryObjectType.GetMethod("TryGetValue", BindingFlags.Public | BindingFlags.Instance)!;
+            MethodInfo tryGetValue = dictionaryType.GetMethod("TryGetValue", DeclaredOnlyLookup)!;
+            PropertyInfo? setter = dictionaryType.GetProperty("Item", DeclaredOnlyLookup);
```

Comment:
(nit, existing)

```suggestion
            PropertyInfo indexerProperty = dictionaryType.GetProperty("Item", DeclaredOnlyLookup)!;
```

1. it doesn't really make sense to call this property `setter`.
2. Since we now know that `dictionaryType` is only `IDictionary` or `Dictionary`, we know that it has an indexer, and the indexer is writable.